### ISSUE TITLE
enhancement: an easier way to obtain an internet access

### ISSUE THREAD

**Issue by legluondunet** [joncampbell123/dosbox-x#320]
Hello,
do you plan to add an easier way of obtain internet access with Dosbox-x?
On Linux it's a pain and you need root rights. I only need an internet access to download files or join game servers. 
PCEM uses SLIRP with NE2000 network card, no additionnal software to install, no parameters to configure on WIndows, it just work.

LGDN

**Comment by joncampbell123** [comment]
I was thinking that two ways could be added:

1. The NE2000 emulation could be updated to emulate TCP and UDP connections as a firewall, complete with an internal DHCP server for Windows to initialize by, which would not require root.

2. I was also thinking that the serial port modem emulation could be updated so that you can "dial out" to a fake ISP and connect out to the internet just like the old dial-up days. DOSBox-X would emulate just enough PPP/SLIP to make Windows 95 think it's a dial-up internet connection.

**Comment by sikthehedgehog** [comment]
> I was also thinking that the serial port modem emulation could be updated so that you can "dial out" to a fake ISP and connect out to the internet just like the old dial-up days. DOSBox-X would emulate just enough PPP/SLIP to make Windows 95 think it's a dial-up internet connection.

Make sure to include a setting that lets it emulate connections as fast as possible instead of the actual speed configured in the emulated hardware (do we want to go back to 56kbps or less? :v)

**Comment by joncampbell123** [comment]
Yes, of course.

I would code it to emulate the dialup connection at the serial port's baud rate, for realism.

Then I would add a dosbox.conf option that allows "locking" the serial port's baud rate at some unnaturally high baud rate that actual hardware would never support. For the impatient, of course. :)

**Comment by joncampbell123** [comment]
SLIP: https://tools.ietf.org/html/rfc1055
CSLIP: https://tools.ietf.org/html/rfc1144#section-3.1
PPP: https://tools.ietf.org/html/rfc1661
PPP CHAP: https://tools.ietf.org/html/rfc1994#section-2.1

**Comment by rderooy** [comment]
I have created a draft wiki page which should help with the NE2000 case. You do not need to give DOSBox-X full root access, you just need to give it the cap_net_raw capability.

https://github.com/Wengier/dosbox-x-wiki/wiki/Guide:-Setting-up-networking-in-DOSBox-X#common-problem-invalid-mac-address

**Comment by rderooy** [comment]
The draft wiki page was moved, here is the new location:
https://github.com/Wengier/dosbox-x-wiki/wiki/Guide:Setting-up-networking-in-DOSBox%E2%80%90X

**Comment by Jookia** [comment]
Slirp support is in master now

**Comment by rderooy** [comment]
Unfortunately, the build is broken.

```
/usr/bin/ld: misc/libmisc.a(ethernet_slirp.o): in function `SlirpEthernetConnection::~SlirpEthernetConnection()':
/home/robert/src/github/dosbox-x/src/misc/ethernet_slirp.cpp:143: undefined reference to `slirp_cleanup'
/usr/bin/ld: misc/libmisc.a(ethernet_slirp.o): in function `SlirpEthernetConnection::Initialize(Section*)':
/home/robert/src/github/dosbox-x/src/misc/ethernet_slirp.cpp:150: undefined reference to `slirp_version_string'
/usr/bin/ld: /home/robert/src/github/dosbox-x/src/misc/ethernet_slirp.cpp:181: undefined reference to `slirp_new'
/usr/bin/ld: misc/libmisc.a(ethernet_slirp.o): in function `SlirpEthernetConnection::GetPackets(std::function<void (unsigned char const*, int)>)':
/home/robert/src/github/dosbox-x/src/misc/ethernet_slirp.cpp:205: undefined reference to `slirp_pollfds_fill'
/usr/bin/ld: /home/robert/src/github/dosbox-x/src/misc/ethernet_slirp.cpp:207: undefined reference to `slirp_pollfds_poll'
/usr/bin/ld: misc/libmisc.a(ethernet_slirp.o): in function `SlirpEthernetConnection::SendPacket(unsigned char const*, int)':
/home/robert/src/github/dosbox-x/src/misc/ethernet_slirp.cpp:196: undefined reference to `slirp_input'
collect2: error: ld returned 1 exit status
```

**Comment by Jookia** [comment]
do you have libslirp-dev installed?

**Comment by rderooy** [comment]
```
$ rpm -qa|grep slirp
libslirp-4.3.1-3.fc33.x86_64
slirp4netns-1.1.8-1.fc33.x86_64
libslirp-devel-4.3.1-3.fc33.x86_64
```

**Comment by rderooy** [comment]
In any case, if the header files are not found, I would expect that it would either not compile in the functionality, or to give some kind or error during the configure step if it is essential.

**Comment by Jookia** [comment]
Could you paste your config.log or something? It should be doing this.

**Comment by rderooy** [comment]
I have attached the log
[config.log](https://github.com/joncampbell123/dosbox-x/files/6100361/config.log)

In any case, configure seems happy enough...
```
checking slirp/libslirp.h usability... yes
checking slirp/libslirp.h presence... yes
checking for slirp/libslirp.h... yes
checking for slirp_input in -lslirp... yes
```

**Comment by Jookia** [comment]
It looks like https://github.com/joncampbell123/dosbox-x/pull/2347/commits/c136cc10524ea8d55a39e2bfe92b8c7ebd564299 may have broke this since it no longer adds slirp's libs or cflags to the build.

**Comment by Wengier** [comment]
@Jookia There have been no changes to the "{libslirp_LIBS}" and "{libslirp_CFLAGS}" parts at all (they are copied as is). On the other hand, "PKG_CHECK_MODULES" is not supported on many environments. What is your suggestion to fix this script?

**Comment by Jookia** [comment]
PKG_CHECK_MODULES sets libslirp_LIBS and libslirp_CFLAGS. Removing it means they're just empty with your changes. How'd it build on your system?

You should be using pkg-config to get the --libs and --cflags output.

**Comment by Wengier** [comment]
@Jookia The original PKG_CHECK_MODULES did not work in my MinGW environment anyway (since it will simply cause the configure to fail), so I modified the configure and makefile manually to make it build. I am certainly not an expert at the configure.ac script.

**Comment by Jookia** [comment]
Ah. It sounds like the configure.ac changes didn't affect the configure/makefile.

**Comment by Wengier** [comment]
@Jookia I always added additional include libs for my MinGW building process, so I modified makefile afterwards as usual. You mentioned the use of pkg-config to get the --libs and --cflags output. Can you please send a PR to make use of it? Thanks!

**Comment by Wengier** [comment]
It was late last night, but I think this will fix the build:

```
    LIBS="$LIBS "`pkg-config slirp --libs`
    CFLAGS="$CFLAGS "`pkg-config slirp --cflags`
    CPPFLAGS="$CPPFLAGS "`pkg-config slirp --cflags`
```

**Comment by Jookia** [comment]
I think that might work!

**Comment by Wengier** [comment]
@Jookia It is confirmed to work on my MinGW environment. Also, along with another script update the MinGW builds without any issues here and without any manual changes needed.

This script code was the only part of my earlier PR that I was not 100% sure, but you said the change looked good at your first glance, so I believed it would work. But that changed at least fixed the configure failure due to the use of PKG_CHECK_MODULES.

**Comment by Jookia** [comment]
If believe the code you wrote just now was the way I originally did it before moving to PKG_CHECK_MODULES.

**Comment by rderooy** [comment]
I can confirm that the code compiles again here.

**Comment by Jookia** [comment]
This might be fixed now.

**Comment by Wengier** [comment]
@Jookia With the latest release it should work now. I tried the Linux Flatpak package myself and the slirp backend certainly works. Thanks for your work!

@legluondunet You can also try the latest Flatpak package, which now supports Internet access via NE2000 and slirp. More information is available from:

https://dosbox-x.com/wiki/Guide%3ASetting-up-networking-in-DOSBox%E2%80%90X

**Comment by legluondunet** [comment]
Hello, I will compile the last git version, test this new feature. and come back to report my tests results.

**Comment by Wengier** [comment]
I hope more people will like this, which definitely makes things easier.

**Comment by brunocastello** [comment]
Any chance of slirp support for macOS hosts?

**Comment by Jookia** [comment]
It should already be supported.

**Comment by brunocastello** [comment]
I tested it, seems that it does not work there yet though.

I've set backend to slirp as per instructions, and added this bit:

[ethernet, slirp]
ipv4_network = 10.0.1.0
ipv4_netmask = 255.255.255.0
ipv4_host = 10.0.1.11
ipv4_nameserver = 10.0.1.2
ipv4_dhcp_start = 10.0.1.0

NE2000 on Windows 98 still complains on boot though. I have even tried running dosbox-x with sudo.

the backend pcap works as it was before when I plug my ethernet to my MacBook Pro.

MacOS Big Sur 11.2.3, Intel MacBook Pro 13" Retina i7 Late 2013, with latest release of dosbox-x

**Comment by Jookia** [comment]
Could you check the log for anything related to Slirp?

**Comment by Wengier** [comment]
When you set backend=slirp and view the DOSBox-X log file, will you find something like "SLIRP: Successfully initialized"? Or "ETHERNET: Unknown ethernet backend: slirp" instead?

**Comment by brunocastello** [comment]
LOG: ETHERNET: Unknown ethernet backend: slirp
LOG: NE2000: Failed to open Ethernet backend slirp

**Comment by Jookia** [comment]
Looks like it didn't compile with Slirp support.

**Comment by brunocastello** [comment]
The build is one of the available releases for download on dosbox-x homepage, Intel Mac version.

**Comment by Jookia** [comment]
I guess the macOS builds don't have Slirp enabled.

**Comment by Wengier** [comment]
@brunocastello Can you build the latest code in your macOS system yourself with the slirp library enabled and see if it works?

**Comment by Jookia** [comment]
Full disclosure, I *think* on some macOS systems homebrew installs the
slirp headers in the wrong directory tree. So checking for that would be
good.

**Comment by brunocastello** [comment]
I am on Big Sur, then how do I get the headers?

**Comment by Jookia** [comment]
Try and use Homebrew to install libslirp/libslirp-dev. Whichever works.
Then see if the build process picks it up.

**Comment by legluondunet** [comment]
Hello,
I can not obtain internet on Windows 95.
I added 

```
[ne2000]
#  ne2000: Enable NE2000 Ethernet emulation.
# nicbase: The base address of the NE2000 board.
#  nicirq: The interrupt it uses. Note serial2 uses IRQ3 as default.
# backend: The backend (pcap or slirp) used for Ethernet emulation.
ne2000  = true
nicbase = 300
nicirq  = 3
backend = slirp

[ethernet, slirp]
#    ipv4_network: The IPv4 network the guest and host services are on.
#    ipv4_netmask: The netmask for the IPv4 network.
#       ipv4_host: The address of the guest on the IPv4 network.
# ipv4_nameserver: The address of the nameserver service provided by the host on the IPv4 network.
# ipv4_dhcp_start: The start address used for DHCP by the host services on the IPv4 network.
ipv4_network    = 10.0.2.0
ipv4_netmask    = 255.255.255.0
ipv4_host       = 10.0.2.2
ipv4_nameserver = 10.0.2.3
ipv4_dhcp_start = 10.0.2.15


[log]
logfile=win95osr2.5-dosbox-x.log
```

to the conf file.
Then in Windows 95 I launched the tool to discover new hardware, it found a network card nec 2000 and install it:

![guest os_008](https://user-images.githubusercontent.com/325405/114133142-de655080-9905-11eb-9102-51e70a073692.png)

but still no internet.
What did I miss?
I have to add tcp/ip protocol?
I have to modify default value in conf file?

Thank you for your help.

**Comment by Jookia** [comment]
What OS are you on? What DOSBox build? Does the log file mention Slirp?

**Comment by legluondunet** [comment]
OS: Ubuntu 20.04
Dosbox build: compiled 8 april 2021
log file: mention slirp --> yes

**Comment by Jookia** [comment]
Huh. Could you paste the Slirp-related lines? At this point it would be
better to open a fresh issue for your issue.

**Comment by legluondunet** [comment]
Ok, I opened a new issue. Good news: it works with WIndows 98.

**Comment by Jookia** [comment]
Quoting from your closed issue:

```
I finally found the problem:
Windows 95 installed NE 2000 network card with IPX/SPX protocol per default.
I deleted it and replaced it by MS TCP/IP protocol and it works now.
Sorry for the noise.
```

Thanks for the investigation! Might be a good idea to add this to documentation somewhere for Win9x installs.

**Comment by brunocastello** [comment]
Turns out I already had libslirp (https://formulae.brew.sh/formula/libslirp) installed, so I just reinstalled it and it fired up a few dependencies updates too. Gonna try to compile a new build of Dosbox with the changes pointed by @Wengier  and see what happens.

**Comment by brunocastello** [comment]
I can confirm that after doing a build with the latest code slirp networking now works for me. Good! Free from the cable now =) Thanks, @Jookia

**Comment by Jookia** [comment]
Does this mean the next build of Slirp for macOS will work?

**Comment by brunocastello** [comment]
Apparently yes. Works well.

However on both latest release and next build I've compiled, I am no longer able to play my 2 FIFA games I had managed to run last night. I am at loss. I can't figure out what is wrong and why the things for these games are not working this time on both builds. But this is another subject and not related to this issue here.

**Comment by Jookia** [comment]
Yeah if you could open an issue about those it'd be great.

**Comment by rderooy** [comment]
The fact that Win95 by default only installs IPX/SPX was already documented on the Wiki. TCP/IP did not become the default until Win98. There are even instructions on the Windows for Workgroups installation steps on the wiki on how to get TCP/IP working.

**Comment by inferpse** [comment]
Hi All,
Thanks for adding slirp support!

So does it mean that next dosbox-x release for ARM/Intel macOS will have slirp support out of the box? 🙂
Latest 0.83.13 release still gives me following output on ARM version of macOS:
```
LOG: ETHERNET: Unknown ethernet backend: slirp
LOG: NE2000: Failed to open Ethernet backend slirp
```

**Comment by Wengier** [comment]
Slirp backend support is particularly useful for Flatpak packages since Flatpak packages do not have pcap backend. With slirp backend available all DOSBox-X builds now support the NE2000 networking feature, with at least one backend, or both. Even if libslirp may not be available on some systems, you can always build on the macOS binary yourself, or perhaps @emendelson will release DOSBox-X-App for macOS featuring slirp backend too.

**Comment by Jookia** [comment]
Does the official CI of macOS have slirp backend?

**Comment by Wengier** [comment]
Actual macOS machines are used for official macOS builds rather than CI. I think the libslirp can be installed using brew:

**brew install libslirp**

Seems to work here. Just in case I may also try to upload macOS builds with libslirp here in case anyone need.

**Comment by brunocastello** [comment]
Please do if you want. I can test this weekend.

**Comment by brunocastello** [comment]
Thanks! I'll give you a feedback tomorrow, it's midnight right now in Brazil... I need a sleep after creating 40+ colorful folder icons for my macOS goodness. I have even created a dosbox folder icon but nothing fancy.

**Comment by brunocastello** [comment]
@Wengier I installed the dependency (fluidsynth), which is totally irrelevant to the case but allowed me to run the binary you sent me. I can confirm that slirp is working on your build for macOS.

I have never needed fluidsynth to run the app, so perhaps it was something you build for yourself? The next public build will not require this, right? 👀

**Comment by Wengier** [comment]
@brunocastello Yes, the macOS builds above were built myself, but in case you don't know yet FluidSynth is an important feature in DOSBox-X too, and it can be installed easily just like how you install libslirp, and once it is installed then it will be always available unless you decide to uninstall it later. I think it is probably a good idea to have a "starter" build (say SDL1 build) without requiring either the FluidSynth and slirp library and a more feature build  (say SDL2 build) with both libraries installed. Hope this helps.

**Comment by brunocastello** [comment]
Ehhh… I don’t think it is a good idea.
SDL1 still produces yellow screen tint for Windows 3.11. I’d say let SDL1 have slirp, and leave fluidsynth for SDL2.

**Comment by inferpse** [comment]
@Wengier it would be great if both SDL release builds could come with libslirp support out of the box since not all people have necessary tooling/knowledge to build the app... And if I get it right this will enable networking capabilities in dosbox-x mac app without the need to install `brew`, `libslirp` or any other additional software?

Will be happy to test ARM build of the app (I'm not using Rosetta)

**Comment by Wengier** [comment]
@inferpse and @brunocastello Since libslirp that comes with brew by default is not a static library, so it may not be easy to have libslirp support out of the box and does not need to install libslirp at the same time. Thus, unless there is a way to solve this it may not be possible to have all macOS release builds come with libslirp support. But the INSTALL.md file will detail the install instructions for running the macOS builds of course.

**Comment by inferpse** [comment]
@Wengier uh, I just wanted to have portable win9x w/ networking which I could start on any mac with zero config and without installing anything else :) 

So maybe it might make sense to have light macOS builds without any dependencies (like today) and fully featured builds for people who could install all necessary dependencies via `brew`? And thanks for manual - I'll check that later!

**Comment by Jookia** [comment]
It might make sense at this point to vendor libslirp.

**Comment by brunocastello** [comment]
I do not mind having to install libslirp via brew. I have to install brew for other work related stuff, anyway. This is A-OK for me.

It's just that I did not need fluidsynth for previous builds, and it was not installed until last weekend to try out this build just to confirm that dosbox-x with slirp works on macOS.

**Comment by Wengier** [comment]
@Jookia I agree that vendoring libslirp will likely solve the issue more or less completely. So does FluidSynth. Internal FluidSynth is already supported in both Visual Studio and MinGW builds, so I would hope it can be supported the same way in macOS builds too, without requiring the external FluidSynth library.

**Comment by Jookia** [comment]
I imagine you'd still have to vendor glib too, which is a headache.

Is there a reason DOSBox-X doesn't just ship with shared libraries?

**Comment by Jookia** [comment]
No, I use Linux.

**Comment by inferpse** [comment]
@Wengier mac hardware is identical for all users and there's no variety of operating systems in Apple's world these days
(especially after M1 chip was introduced). So I guess macOS builds might come bundled with all necessary libraries out of the box.

When I did `brew install libslirp` it did not build or install anything from the sources - it simply downloaded precompiled binary bottle (and the same was for `glib`)

In any case I think you have @brunocastello and myself available to test some builds on Intel and ARM-based macs)

**Comment by lavadrop** [comment]
> @Wengier mac hardware is identical for all users and there's no variety of operating systems in Apple's world these days
> (especially after M1 chip was introduced). So I guess macOS builds might come bundled with all necessary libraries out of the box.
> 
> When I did `brew install libslirp` it did not build or install anything from the sources - it simply downloaded precompiled binary bottle (and the same was for `glib`)
> 
> In any case I think you have @brunocastello and myself available to test some builds on Intel and ARM-based macs)

Yeah, not really. The default SDL2 download is missing the referenced libslirp library and crashes on my 10.15.7 macos install. The built shared by @Wengier however, is missing libfluidsynth and crashes too.
Can we get a build with both included?

**Comment by Wengier** [comment]
@lavadrop As mentioned in the installation instructions, please install both libraries before running either build (for now). The steps are very simple:

```
brew install fluid-synth
brew install libslirp
```

Hope this helps.

**Comment by brunocastello** [comment]
Just my two cents, that regarding dosbox-x, my focus is on trying the SDL1 builds, because SDL2 builds still hasn’t fixed the yellow tinted windows desktops on win 3.x and 9x, and it was documented that the OpenGlide method currently only works with SDL1 builds. I don’t mind having to do a brew install of a few required libraries, for stuff I need from dosbox-x, as long as support is still there for us. I am familiar with brew, mostly because of my previous career as a web frontend dev.

**Comment by Wengier** [comment]
@brunocastello The issue regarding yellow tinted Windows desktop only happens with specific video drivers, and changing the drivers will fix the issue. As for Glide support, I think @grapeli has found a way to workaround the main issue with SDL2 builds via of sdl12-compat (see #2126).

**Comment by brunocastello** [comment]
> The issue regarding yellow tinted Windows desktop only happens with specific video drivers, and changing the drivers will fix the issue. 

Was it documented on dosbox-x wiki for windows installs, along with the required drivers for it? I don't remember seeing that before, so excuse me in advance if it was documented later.

> As for Glide support, I think @grapeli has found a way to workaround the main issue with SDL2 builds via [sdl12-compat](https://github.com/libsdl-org/sdl12-compat) (see #2126).

He's apparently using linux. I think I have to talk with @almeath to see if he managed to compile OpenGlide with SDL2 for macOS.

**Comment by Wengier** [comment]
@brunocastello With the video driver I used (probably the default one) the issue does not occur, so I cannot really tell exactly which drivers cause issues and which do not. But you can certainly try a different video driver and see the result, as all I can tell is that the issue does not occur with some video drivers.

**Comment by almeath** [comment]
> He's apparently using linux. I think I have to talk with @almeath to see if he managed to compile OpenGlide with SDL2 for macOS.

I have tried to get it working with SDL2 but DOSBox-X segfaults when I launch the glide enabled game in full screen, or shows a black screen with only sound but no graphics when launching in windowed mode, so no luck so far. I test with Lands of Lore 2, which is working fine with SDL1 (both DOSBox-X and SVN).
